test/parallel_api/numeric/numeric.ops/in_place_scan.pass.cpp

```diff
@@ -19,64 +19,570 @@
+#if TEST_DPCPP_BACKEND_PRESENT
+        !oneapi::dpl::__internal::__is_hetero_execution_policy<typename ::std::decay<Policy>::type>::value &&
+#endif
+            is_base_of_iterator_category<::std::random_access_iterator_tag, Iterator1>::value,
+        void>::type
+    operator()(Policy&& exec, Iterator1 keys_first, Iterator1 keys_last, Size n)
+    {
+        typedef typename ::std::iterator_traits<Iterator1>::value_type KeyT;
+
+        // call algorithm with no optional arguments
+        initialize_data(keys_first, n);
+
+        // copy source host keys state into temp buf
+        typedef typename ::std::iterator_traits<Iterator1>::value_type ValT;
+        const std::vector<ValT> source_host_keys_state(keys_first, keys_first + n);
+
+        auto res1 = oneapi::dpl::exclusive_scan(exec, keys_first, keys_last, keys_first, kInitValue);
+        check_values(source_host_keys_state, keys_first, n);
+
+        // call algorithm with equality comparator
+        initialize_data(keys_first, n);
+        auto res2 = oneapi::dpl::exclusive_scan(exec, keys_first, keys_last, keys_first, DefaultBinaryOp<KeyT>());
+        check_values(source_host_keys_state, keys_first, n, DefaultBinaryOp<KeyT>());
+
+        // call algorithm with addition operator
+        initialize_data(keys_first, n);
+        auto res3 = oneapi::dpl::exclusive_scan(exec, keys_first, keys_last, keys_first, kInitValue, BinaryOperation());
+        check_values(source_host_keys_state, keys_first, n, BinaryOperation());
+    }
+
+    // specialization for non-random_access iterators
+    template <typename Policy, typename Iterator1, typename Size>
+    typename ::std::enable_if<!is_base_of_iterator_category<::std::random_access_iterator_tag, Iterator1>::value, void>::type
+    operator()(Policy&& exec, Iterator1 keys_first, Iterator1 keys_last, Size n)
+    {
+    }
+};
+
+DEFINE_TEST_1(test_exclusive_scan_non_inplace, BinaryOperation)
+{
+    DEFINE_TEST_CONSTRUCTOR(test_exclusive_scan_non_inplace)
+
+    const int kInitValue = 2;
+
+    // TODO: replace data generation with random data and update check to compare result to
+    // the result of a serial implementation of the algorithm
+    template <typename Iterator1, typename Size>
+    void
+    initialize_data(Iterator1 host_keys, Size n)
+    {
+        const Size kStartVal = 1;
+        
+        for (Size i = 0; i != n; ++i)
+            host_keys[i] = kStartVal + i;
+    }
+
+    // Required to pass source_host_keys_state param by copy for safe state of this param on caller side
+    template <typename ValT, typename Iterator1, typename Size,
+              typename BinaryOperationCheck = oneapi::dpl::__internal::__pstl_plus>
+    void check_values(std::vector<ValT> source_host_keys_state, Iterator1 host_keys, Size n,
+                      BinaryOperationCheck op = BinaryOperationCheck())
+    {
+        const auto source_host_keys_state_copy = source_host_keys_state;
+
+        if (source_host_keys_state.empty())
```

Comment:
Is it really situation when source_host_keys_state may be empty?
Probably, it makes sense to add just assert here?

Reply:
Fixed